### PR TITLE
eth/tracers: fix gasUsed for native and JS tracers

### LINKED ISSUE #26018: gasUsed incorrect when using `debug_traceBlockByHash`
#### System information

Geth version: `Geth 1.10.24`
CL client & version: prysm:3.1.1
OS & Version: Linux
Commit hash : 1.10.24 release branch

#### Expected behaviour
When using `debug_traceBlockByHash` with `callTracer` I expect the `gasUsed` to be the `gasUsed` by the transaction. (https://geth.ethereum.org/docs/dapp/builtin-tracers#calltracer) documentation

i.e. on `Goerli Testnet` 

```
curl --request POST \
  --url http://127.0.0.1:8546/ \
  --header 'content-type: application/json' \
  --data '{"method":"debug_traceBlockByHash","params":["0x436ec05f6768296d48150241a1217000f2d7a4cfd8e36803150f2d5219ebab7b", {
	"tracer": "callTracer"
}], "jsonrpc":"2.0", "id": 0}'
```
(https://goerli.etherscan.io/tx/0x5bd8217569235b62252bc5c5063d1e6e9ee9e450ae7370c2d9c349ca1692ef8a) is an example

#### Actual behaviour
It outputs the wrong gasUsed. For that specific txn: gas used is: `0x24ca6`
Out put is:

```
		{
			"result": {
				"type": "CALL",
				"from": "0x7afe817effeaffd965f6336137c027a15dbde4c7",
				"to": "0x120b1b0a44d680524419f76e1a41976b6ea0add5",
				"value": "0x11c37937e080000",
				"gas": "0x46790",
				"gasUsed": "0x24ca6",
				"input": "0xcf11d58d00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001",
				"output": "0x",
				"calls": [
					{
						"type": "CALL",
						"from": "0x120b1b0a44d680524419f76e1a41976b6ea0add5",
						"to": "0x727b51b3cebe69eaabdcfd5a2db0c521ba308112",
						"value": "0x11c37937e080000",
						"gas": "0x8fc",
						"gasUsed": "0x0",
						"input": "0x",
						"output": "0x"
					}
				]
			}
		}
```

if you run `eth_getTransactionReceipt` on the same tx hash: `0x5bd8217569235b62252bc5c5063d1e6e9ee9e450ae7370c2d9c349ca1692ef8a` 
gas used is: `0x2a006` 
```

{
	"jsonrpc": "2.0",
	"id": 0,
	"result": {
		"blockHash": "0x8b973f65653ba08a81fe728ede379cc748ae5ee959b8925ef637a35162bdb0ca",
		"blockNumber": "0x764710",
		"contractAddress": null,
		"cumulativeGasUsed": "0x2f20e",
		"effectiveGasPrice": "0x113abe6400",
		"from": "0x7afe817effeaffd965f6336137c027a15dbde4c7",
		"gasUsed": "0x2a006",
		"logs": [
			{
				"address": "0x120b1b0a44d680524419f76e1a41976b6ea0add5",
				"topics": [
					"0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
					"0x0000000000000000000000000000000000000000000000000000000000000000",
					"0x0000000000000000000000007afe817effeaffd965f6336137c027a15dbde4c7",
					"0x000000000000000000000000000000000000000000000000000000000000002b"
				],
				"data": "0x",
				"blockNumber": "0x764710",
				"transactionHash": "0x5bd8217569235b62252bc5c5063d1e6e9ee9e450ae7370c2d9c349ca1692ef8a",
				"transactionIndex": "0x1",
				"blockHash": "0x8b973f65653ba08a81fe728ede379cc748ae5ee959b8925ef637a35162bdb0ca",
				"logIndex": "0x0",
				"removed": false
			}
		],
		"logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000800008000000000000000000000000000000000000000000000000020000000000000000000800000000000000080000000010002000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000200000000000000000000000000000000002000000100000000000000000000000000000000000000000000020000000000000000000000008000000000000000000000000000000000000000002",
		"status": "0x1",
		"to": "0x120b1b0a44d680524419f76e1a41976b6ea0add5",
		"transactionHash": "0x5bd8217569235b62252bc5c5063d1e6e9ee9e450ae7370c2d9c349ca1692ef8a",
		"transactionIndex": "0x1",
		"type": "0x2"
	}
}
```
#### Steps to reproduce the behaviour
1. Pick any block hash you want on the network. In this example I'm using goerli with: `0x436ec05f6768296d48150241a1217000f2d7a4cfd8e36803150f2d5219ebab7b`
2. Run debug command:

```
{"method":"debug_traceBlockByHash","params":["0x436ec05f6768296d48150241a1217000f2d7a4cfd8e36803150f2d5219ebab7b", {
	"tracer": "callTracer"
}], "jsonrpc":"2.0", "id": 0}
```

3. The gasUsed is incorrect for every transaction. If you look at the second one, you will see. 
```
				"gasUsed": "0x24ca6",

```
4. Now if you run 
```
curl --request POST \
  --url http://0.0.0.0:8546/ \
  --header 'Authorization: Bearer 0xc9cdf7d459216ebd1218521b7df4946f28bf547d6d97effb6e79ab1550ca2dd3' \
  --header 'Content-Type: application/json' \
  --data '{
  "jsonrpc" : "2.0",
  "method" : "eth_getTransactionReceipt",
  "params" : ["0x5bd8217569235b62252bc5c5063d1e6e9ee9e450ae7370c2d9c349ca1692ef8a"],
  "id" : 0
}
'
```
the gas used is different.

#### Backtrace

````
[backtrace]
````

When submitting logs: please submit them as text and not screenshots.

### PR DESCRIPTION
**Pull request by s1na**
Fixes the gasUsed value in the following contexts:

- Native and legacy callTracer's top frame gasUsed value. I think this is tx-level gas accounting affecting a call frame.
- The gasUsed value returned to JS tracers on `result`
- Legacy prestateTracer would have had wrong sender balance in case there were refunds (i think)

I like to in future change `CaptureEnd` to drop `gasUsed` and `timeDuration` both of which are kinda useless.

### Backwards-compatibility note

The field `ctx['intrinsicGas']` in `result(ctx)` will be dropped from the JS tracer. The field `ctx['gasUsed']` in `result(ctx)` will change its value. It previously did not account for gas refunds. Hence if there are refunds the `gasUsed` value will be lower than before.

Likewise the `gasUsed` field of the top call frame of the `callTracer` will change. The value did not previously account for the intrinsic gas (e.g. 21000 tx cost) and gas refunds.

Fixes #26018

### CODE REVIEW

File: eth/tracers/js/tracer_test.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func TestTracer(t *testing.T) {
 			want: `["PUSH1","PUSH1","STOP"]`,
 		}, { // tests intrinsic gas
 			code: "{depths: [], step: function() {}, fault: function() {}, result: function(ctx) { return ctx.gasPrice+'.'+ctx.gasUsed+'.'+ctx.intrinsicGas; }}",
-			want: `"100000.6.21000"`,
+			want: `"100000.30994.21000"`,
```
Comment (holiman) [inline eth/tracers/js/tracer_test.go:133]: So the execution of the three opcodes `PUSH1, PUSH1, STOP` burns `30994` gas? How is that possible?

Comment (s1na) [reply eth/tracers/js/tracer_test.go]: Good point haha, there was an issue in `runTrace`. Fixed.

**Comment by holiman** [comment]
If we have these entities;
- `exec_gas` - gas burnt during execution
- `intrinsic_gas` - the intrinsic cost
- `refund`

Does this change mean that `gasUsed` is `exec_gas + intrinsic_gas`, and previously was only `exec_gas` ? 
And, if so, exactly what tracers are affected? 
This sounds like a backwards-incomaptible change which we should make sure to mention in the release-docs later on.

**Comment by s1na** [comment]
Yes, to be more precise it's `exec_gas + intrinsic_gas - refund`. All JS tracers that make use of this field are affected. The built-in call and legacyCall tracers have their gasUsed values changed as well. I agree with announcing it in release docs I think people should be aware it's changing under their feet.

**Comment by holiman** [comment]
Please write up, in the PR description, a blurb which describes this change and could be lifted into the future release-notes, complete with warning about how things may malfunction/differ. 


Also, I'm curious. Was there any error here, or was it simply a case of lack of clarity about what was counted in the `usedGas` field?

**Comment by s1na** [comment]
> Also, I'm curious. Was there any error here, or was it simply a case of lack of clarity about what was counted in the usedGas field?

Historically complete information about gas usage was not exposed to the tracers so they had an incomplete gasUsed value. Now that we've added the `TxStart` and `TxEnd` hooks we can integrate intrinsic and refund gas.

**Comment by s1na** [comment]
Thinking about it, I'm gonna go ahead and remove the `ctx['intrinsicGas'] field. It was introduced in https://github.com/ethereum/go-ethereum/pull/22038 to fix this exact issue. But if we go ahead with this PR then it won't be needed.

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**